Potential bug running against Scala.js

When trying to run `scalafmt` on the Scala.js code base as seen here:
https://github.com/scala-js/scala-js/pull/4522

Config: https://github.com/scala-js/scala-js/pull/4522/commits/f02a0886a203f338705b1131bf59da4a6567d417

## Steps

Given code like this:
https://github.com/ekrich/scala-js/tree/topic/scalafmt

When I run scalafmt like this:

```bash
scripts/scalafmt --test 2> diff.txt
```

## Problem

Scalafmt formats code like this:
```scala
org.scalafmt.cli.FailedToFormat: /Users/eric/workspace/scala-js/linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/NameGen.scala
Caused by: org.scalafmt.Error$PreciseIncomplete: /Users/eric/workspace/scala-js/linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/NameGen.scala:72: error: Unable to format file due to bug in scalafmt
    cache.getOrElseUpdate(name, {
                              ^
	at org.scalafmt.Scalafmt$.$anonfun$doFormatOne$6(Scalafmt.scala:133)
	at scala.util.Success.flatMap(Try.scala:258)
	at org.scalafmt.Scalafmt$.$anonfun$doFormatOne$3(Scalafmt.scala:128)
	at scala.meta.parsers.Parsed.fold(Errors.scala:12)
	at scala.meta.parsers.Parsed.fold$(Errors.scala:11)
	at scala.meta.parsers.Parsed$Success.fold(Errors.scala:26)
	at org.scalafmt.Scalafmt$.doFormatOne(Scalafmt.scala:124)
	at org.scalafmt.Scalafmt$.doFormat(Scalafmt.scala:102)
	at org.scalafmt.Scalafmt$.formatCode(Scalafmt.scala:71)
	at org.scalafmt.cli.ScalafmtCoreRunner$.unsafeHandleFile(ScalafmtCoreRunner.scala:78)
	at org.scalafmt.cli.ScalafmtCoreRunner$.handleFile(ScalafmtCoreRunner.scala:60)
	at org.scalafmt.cli.ScalafmtCoreRunner$.$anonfun$run$5(ScalafmtCoreRunner.scala:40)
	at org.scalafmt.cli.ScalafmtCoreRunner$.$anonfun$run$5$adapted(ScalafmtCoreRunner.scala:39)
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:563)
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:561)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1288)
	at scala.collection.parallel.ParIterableLike$Foreach.leaf(ParIterableLike.scala:938)
	at scala.collection.parallel.Task.$anonfun$tryLeaf$1(Tasks.scala:52)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.util.control.Breaks$$anon$1.catchBreak(Breaks.scala:97)
	at scala.collection.parallel.Task.tryLeaf(Tasks.scala:55)
	at scala.collection.parallel.Task.tryLeaf$(Tasks.scala:49)
	at scala.collection.parallel.ParIterableLike$Foreach.tryLeaf(ParIterableLike.scala:935)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$AWSTWrappedTask.internal(Tasks.scala:169)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$AWSTWrappedTask.internal$(Tasks.scala:156)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$AWSFJTWrappedTask.internal(Tasks.scala:304)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$AWSTWrappedTask.compute(Tasks.scala:149)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$AWSTWrappedTask.compute$(Tasks.scala:148)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$AWSFJTWrappedTask.compute(Tasks.scala:304)
	at java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:189)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

## Expectation

No error. There are also quite a few "Search state exploded ..." even with the setting at `100,000`. Not sure what is a reasonable number for that or if the files can be formatted.

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/BestFirstSearch.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/BestFirstSearch.scala
@@ -154,7 +154,9 @@ private class BestFirstSearch private (
             else Q.clear()
         }
 
-        val blockClose = shouldRecurseOnBlock(splitToken, stop)
+        val blockClose =
+          if (start.eq(curr) && 0 != maxCost) None
+          else shouldRecurseOnBlock(splitToken, stop)
         if (blockClose.nonEmpty)
           blockClose.foreach { end =>
             shortestPathMemo(curr, end, depth + 1, maxCost)

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/BestFirstSearch.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/BestFirstSearch.scala
@@ -60,15 +60,13 @@ private class BestFirstSearch private (
       // TODO(olafur) document why/how this optimization works.
       !best.get(curr.depth).exists(_.alwaysBetter(curr))
 
-  def shouldRecurseOnBlock(
-      ft: FormatToken,
-      stop: Token,
+  def shouldRecurseOnBlock(ft: FormatToken, stop: Token)(implicit
       style: ScalafmtConfig
   ): Option[Token] =
     if (!recurseOnBlocks || !isInsideNoOptZone(ft)) None
     else {
       val left = tokens(ft, -1)
-      val closeOpt = formatOps.getEndOfBlock(left, false)(style)
+      val closeOpt = formatOps.getEndOfBlock(left, false)
       closeOpt.filter(close =>
         // Block must span at least 3 lines to be worth recursing.
         close != stop &&
@@ -156,10 +154,12 @@ private class BestFirstSearch private (
             else Q.clear()
         }
 
-        val blockClose = shouldRecurseOnBlock(splitToken, stop, style)
+        val blockClose = shouldRecurseOnBlock(splitToken, stop)
         if (blockClose.nonEmpty)
-          shortestPathMemo(curr, blockClose.get, depth + 1, maxCost)
-            .foreach(Q.enqueue(_))
+          blockClose.foreach { end =>
+            shortestPathMemo(curr, end, depth + 1, maxCost)
+              .foreach(Q.enqueue(_))
+          }
         else if (
           escapeInPathologicalCases &&
           routes(curr.depth).lengthCompare(1) != 0 &&

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/BestFirstSearch.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/BestFirstSearch.scala
@@ -150,8 +150,7 @@ private class BestFirstSearch private (
             dequeueSpots.contains(tokenHash) &&
             (depth > 0 || !isInsideNoOptZone(splitToken))
           )
-            if (depth == 0) addGeneration
-            else Q.clear()
+            addGeneration
         }
 
         val blockClose =

